File: binary-tree/basicBT.py
class Node():

    def __init__(self, data):
        self.left = None
        self.right = None
        self.data = data

    def insert(self, data):
        if data < self.data:
            if self.left is None:
                self.left = Node(data)
            else:
                self.left.insert(data)
        if data > self.data:
            if self.right is None:
                self.right = Node(data)
            else:
                self.right.insert(data)

    def lookup(self, data, parent=None):
        if data < self.data:
            if self.left is None:
                return None, None
            return self.left.lookup(data, self)
        elif data > self.data:
            if self.right is None:
                return None, None
            return self.right.lookup(data, self)
        else:
            return self.data, parent


def main():
    r = Node(8)
    r.insert(3)
    r.insert(10)
    r.insert(1)

File: binary-tree/basicBT.go
package main

import (
	. "fmt"
)

type Node struct {
	ParentNode   *Node
	Data         int
	ChildrenNode [2]*Node
}

func (n *Node) Insert(data int) {
	nodeTemp := Node{Data: data, ParentNode: n}
	if data < n.Data {
		if n.ChildrenNode[0] != nil {
			n.ChildrenNode[0].Insert(data)
		} else {
			n.ChildrenNode[0] = &nodeTemp
		}
	} else if data > n.Data {
		if n.ChildrenNode[1] != nil {
			n.ChildrenNode[1].Insert(data)
		} else {
			n.ChildrenNode[1] = &nodeTemp
		}
	}
}

func (n *Node) Lookup(data int) (*Node, *Node) {
	var rn, p *Node
	if data < n.Data {
		if n.ChildrenNode[0] == nil {
			rn = nil
			p = n
		} else {
			rn, p = n.ChildrenNode[0].Lookup(data)
		}
	} else if data > n.Data {
		if n.ChildrenNode[1] == nil {
			rn = nil
			p = n
		} else {
			rn, p = n.ChildrenNode[1].Lookup(data)
		}
	} else {
		rn = n
		p = n.ParentNode
	}
	return rn, p
}

func main() {
	root := Node{Data: 8}

	root.Insert(3)
	root.Insert(10)
	root.Insert(1)
	root.Insert(6)
	root.Insert(4)
	root.Insert(7)
	root.Insert(14)
	root.Insert(13)

	Println(root.Data)
	Println(root.ChildrenNode[0])
	Println(root.Lookup(4))
}
